### DISCUSSION TITLE
EdgeX APIs, CLI and GUI usage

### DISCUSSION THREAD

**Discussion by JamesKButcher** [edgexfoundry/community#29] [General]
Hi EdgeX Community!

I’m looking for some responses from the users of EdgeX as to how they interact with EdgeX please…

Obviously we have well defined APIs for controlling and interfacing with EdgeX. But we also have a Command Line Interface (CLI) and a Graphical User Interface (GUI).

To help us better understand usage and plan areas of focus, particularly as we work towards version 3.0, I’d like to hear whether people generally use all of the above (i.e. CLI and/or GUI) or whether they focus on using just the APIs. Please let me know any feedback below...

Thanks all!
James Butcher
EdgeX TSC Chair

**Comment by sounak-1989** [comment]
In my project I use only the APIs. I don't even run the UI application.

**Comment by JamesKButcher** [comment]
Thanks @sounak-1989. 

I was hoping for a few more replies so I added an easy poll here: https://github.com/orgs/edgexfoundry/discussions/34 
Hopefully just 2 seconds for folks to click one of the options there

**Comment by ed-lisle** [comment]
All tools are great... and we plan to utilize all the above to accelerate our services.

The Edgex Console UI is nice when learning and engaging in development phases of the stack. I do not see this being used by our customer base and end users but more so for our own development and testing. And maybe for support services.

API’s are great and necessary when developing our microservices. This is the way to go and that is what we are doing.

Some activities when performing provisioning, setup, replication, … can be efficiently performed with scripts using the command line tools. Especially to fast tract field testing deployments before our microservices have matured. Otherwise, one would have to implement the API’s for straight forward tasks that can be performed by a simple script.

Totally understand the engineering commitment. I just assumed that EdgeX is matured enough to justify having a lot of bells and whistles to accelerate adoption. I hope this helps as I'm a big supporter of more tools not less.

**Reply by ed-lisle** [reply]
I do have a feature request for the edgex-cli... since it support the -json option to extract various components of EdgeX it would be nice to have the reverse function to pipe the entire raw json file back into the stack. The reason is if we need a fast method to tune or adjust various items in the stack. That we could perform these tests and then factor them in our future microservice development.

An example is that I needed to replicate missing sensor variables and commands in a device profile to model and test various usage models. So using a script to replicate this on various edgex instances can help accelerate my field test deployment before committing to code development.

This is just one example as I learn how powerful the edgex-cli can be. Although as I mentioned above, command line support can be highly useful in other ways.
